pkg/sql/temporary_tables_deletion.go

```diff
@@ -0,0 +1,226 @@
+	"github.com/cockroachdb/cockroach/pkg/security"
+	"github.com/cockroachdb/cockroach/pkg/settings/cluster"
+	"github.com/cockroachdb/cockroach/pkg/sql/sem/tree"
+	"github.com/cockroachdb/cockroach/pkg/util/log"
+	"github.com/cockroachdb/cockroach/pkg/util/stop"
+	"github.com/cockroachdb/cockroach/pkg/util/uint128"
+)
+
+// deleteTempTablesResumer implements the jobs.Resumer interface.
+type deleteTempTablesResumer struct {
+	job *jobs.Job
+}
+
+var _ jobs.Resumer = &deleteTempTablesResumer{}
+
+// defaultCleanupInterval is the interval of time after which a background job
+// runs to cleanup temporary tables that weren't deleted at session exit.
+var defaultCleanupInterval = 30 * time.Minute
+
+// Resume is part of the jobs.Resumer interface.
+func (r *deleteTempTablesResumer) Resume(
+	ctx context.Context, phs interface{}, resultsCh chan<- tree.Datums,
+) error {
+	p := phs.(*planner)
+	p.txn = client.NewTxn(ctx, p.ExecCfg().DB, 0, client.RootTxn)
+
+	// There should only be one temp table deletion job at any point.
+	if err := checkRunningTempTablesJob(ctx, r.job, p); err != nil {
+		return err
+	}
+
+	req := p.makeSessionsRequest(ctx)
+	response, err := p.extendedEvalCtx.StatusServer.ListSessions(ctx, &req)
+	if err != nil {
+		return err
+	}
+	activeSessions := make(map[uint128.Uint128]bool)
+	for _, session := range response.Sessions {
+		activeSessions[uint128.FromBytes(session.ID)] = true
+	}
+	dbIDs, err := GetAllDatabaseDescriptorIDs(ctx, p.txn)
+	if err != nil {
+		return err
+	}
+	for _, dbID := range dbIDs {
+		schemaNames, err := p.GetAllSchemaNames(ctx, dbID, p.txn)
+		if err != nil {
+			return err
+		}
+		for _, scName := range schemaNames {
+			isTempSchema, sessionID, err := temporarySchemaSessionID(scName)
+			if err != nil {
+				return err
+			}
+			if isTempSchema && !activeSessions[sessionID.Uint128] {
+				if err := cleanupSessionTempObjects(ctx, p, sessionID); err != nil {
+					return err
+				}
+			}
+		}
+	}
+	return nil
+}
+
+// OnFailOrCancel is part of the jobs.Resumer interface.
+func (r *deleteTempTablesResumer) OnFailOrCancel(context.Context, *client.Txn) error {
+	return nil
+}
+
+// OnSuccess is part of the jobs.Resumer interface.
+func (r *deleteTempTablesResumer) OnSuccess(ctx context.Context, _ *client.Txn) error { return nil }
+
+// OnTerminal is part of the jobs.Resumer interface.
+func (r *deleteTempTablesResumer) OnTerminal(context.Context, jobs.Status, chan<- tree.Datums) {}
+
+type concurrentDeleteTempTablesError struct{}
+
+var _ error = concurrentDeleteTempTablesError{}
+
+func (concurrentDeleteTempTablesError) Error() string {
+	return "another delete temp tables job is already running"
+}
+
+// ConcurrentDeleteTempTablesError is reported when two delete temp tables jobs
+// are issued concurrently. This is a sentinel error.
+var ConcurrentDeleteTempTablesError error = concurrentDeleteTempTablesError{}
+
+func checkRunningTempTablesJob(ctx context.Context, job *jobs.Job, p *planner) error {
```

Comment:
This function is pretty much copy-pasted from `checkRunningJobs`. Could you refactor to share more code?